core/trie/bitarray.go

```diff
@@ -0,0 +1,397 @@
+// It uses a little endian representation to do bitwise operations of the words efficiently.
+// For example, if len is 10, it means that the 2^9, 2^8, ..., 2^0 bits are used.
+// The max length is 255 bits (uint8), because our use case only need up to 251 bits for a given trie key.
+// Although words can be used to represent 256 bits, we don't want to add an additional byte for the length.
+type BitArray struct {
+	len   uint8     // number of used bits
+	words [4]uint64 // little endian (i.e. words[0] is the least significant)
+}
+
+func (b *BitArray) Felt() *felt.Felt {
+	bs := b.Bytes()
+	return new(felt.Felt).SetBytes(bs[:])
+}
+
+func (b *BitArray) Len() uint8 {
+	return b.len
+}
+
+// Bytes returns the bytes representation of the bit array in big endian format
+//
+//nolint:mnd
+func (b *BitArray) Bytes() [32]byte {
+	var res [32]byte
+
+	switch {
+	case b.len == 0:
+		// all zeros
+		return res
+	case b.len >= 192:
+		// Create mask for top word: keeps only valid bits above 192
+		// e.g., if len=200, keeps lowest 8 bits (200-192)
+		mask := maxUint64 >> (256 - uint16(b.len))
+		binary.BigEndian.PutUint64(res[0:8], b.words[3]&mask)
+		binary.BigEndian.PutUint64(res[8:16], b.words[2])
+		binary.BigEndian.PutUint64(res[16:24], b.words[1])
+		binary.BigEndian.PutUint64(res[24:32], b.words[0])
+	case b.len >= 128:
+		// Mask for bits 128-191: keeps only valid bits above 128
+		// e.g., if len=150, keeps lowest 22 bits (150-128)
+		mask := maxUint64 >> (192 - b.len)
+		binary.BigEndian.PutUint64(res[8:16], b.words[2]&mask)
+		binary.BigEndian.PutUint64(res[16:24], b.words[1])
+		binary.BigEndian.PutUint64(res[24:32], b.words[0])
+	case b.len >= 64:
+		// You get the idea
+		mask := maxUint64 >> (128 - b.len)
+		binary.BigEndian.PutUint64(res[16:24], b.words[1]&mask)
+		binary.BigEndian.PutUint64(res[24:32], b.words[0])
+	default:
+		mask := maxUint64 >> (64 - b.len)
+		binary.BigEndian.PutUint64(res[24:32], b.words[0]&mask)
+	}
+
+	return res
+}
+
+// EqualMSBs checks if two bit arrays share the same most significant bits, where the length of
+// the check is determined by the shorter array. Returns true if either array has
+// length 0, or if the first min(b.len, x.len) MSBs are identical.
+//
+// For example:
+//
+//	a = 1101 (len=4)
+//	b = 11010111 (len=8)
+//	a.EqualMSBs(b) = true  // First 4 MSBs match
+//
+//	a = 1100 (len=4)
+//	b = 1101 (len=4)
+//	a.EqualMSBs(b) = false // All bits compared, not equal
+//
+//	a = 1100 (len=4)
+//	b = [] (len=0)
+//	a.EqualMSBs(b) = true  // Zero length is always a prefix match
+func (b *BitArray) EqualMSBs(x *BitArray) bool {
+	if b.len == x.len {
+		return b.Equal(x)
+	}
+
+	if b.len == 0 || x.len == 0 {
+		return true
+	}
+
+	var long, short *BitArray
+
+	long, short = b, x
+	if b.len < x.len {
+		long, short = x, b
+	}
+
+	return long.Rsh(long, long.len-short.len).Equal(short)
+}
+
+// Truncate sets b to the first 'length' bits of x (starting from the least significant bit).
```

Comment:
```
// Truncate sets the bit array ...
```
Suggestion to be inline with the previous functions which never mention `b`.

I would also omit the name of the function from the description since it can be infered from the function name. Smth like:

```
// Sets the bit array ...
```

Reply:
I've seen that the code style starts always with the function name, so I think is better to maintain it